File: day11/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	grid := [][]byte{}
	grid2 := [][]byte{}
	tmpGrid := [][]byte{}
	tmpGrid2 := [][]byte{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		k := []byte(scanner.Text())
		grid = append(grid, k)
		tmpGrid = append(tmpGrid, make([]byte, len(k)))
		k2 := []byte(scanner.Text())
		grid2 = append(grid2, k2)
		tmpGrid2 = append(tmpGrid2, make([]byte, len(k2)))
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	untilStable(false, grid, tmpGrid)
	fmt.Println("Part 1:", countSeats(grid))

	untilStable(true, grid2, tmpGrid2)
	fmt.Println("Part 2:", countSeats(grid2))
}

func countSeats(grid [][]byte) int {
	count := 0
	for _, i := range grid {
		for _, j := range i {
			if j == '#' {
				count++
			}
		}
	}
	return count
}

func untilStable(mode bool, grid, tmpGrid [][]byte) {
	for {
		if !nextGrid(mode, grid, tmpGrid) {
			break
		}
		tmp := grid
		grid = tmpGrid
		tmpGrid = tmp
	}
}

func nextGrid(mode bool, grid [][]byte, next [][]byte) bool {
	change := false
	for i, r := range grid {
		for j, c := range r {
			var k byte
			if mode {
				k = nextSeat2(i, j, grid)
			} else {
				k = nextSeat(i, j, grid)
			}
			if k != c {
				change = true
			}
			next[i][j] = k
		}
	}
	return change
}

func nextSeat2(i, j int, grid [][]byte) byte {
	c := grid[i][j]
	s1 := seatVal2(i, j, -1, 0, grid)
	s2 := seatVal2(i, j, -1, 1, grid)
	s3 := seatVal2(i, j, 0, 1, grid)
	s4 := seatVal2(i, j, 1, 1, grid)
	s5 := seatVal2(i, j, 1, 0, grid)
	s6 := seatVal2(i, j, 1, -1, grid)
	s7 := seatVal2(i, j, 0, -1, grid)
	s8 := seatVal2(i, j, -1, -1, grid)
	k := s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8
	if c == 'L' && k == 0 {
		return '#'
	}
	if c == '#' && k >= 5 {
		return 'L'
	}
	return c
}

func seatVal2(i, j int, di, dj int, grid [][]byte) int {
	i += di
	j += dj
	for inBounds(i, 0, len(grid)-1) && inBounds(j, 0, len(grid[0])-1) {
		if grid[i][j] == 'L' {
			return 0
		}
		if grid[i][j] == '#' {
			return 1
		}
		i += di
		j += dj
	}
	return 0
}

func nextSeat(i, j int, grid [][]byte) byte {
	c := grid[i][j]
	s1 := seatVal(i-1, j, grid)
	s2 := seatVal(i-1, j+1, grid)
	s3 := seatVal(i, j+1, grid)
	s4 := seatVal(i+1, j+1, grid)
	s5 := seatVal(i+1, j, grid)
	s6 := seatVal(i+1, j-1, grid)
	s7 := seatVal(i, j-1, grid)
	s8 := seatVal(i-1, j-1, grid)
	k := s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8
	if c == 'L' && k == 0 {
		return '#'
	}
	if c == '#' && k >= 4 {
		return 'L'
	}
	return c
}

func seatVal(i, j int, grid [][]byte) int {
	if !inBounds(i, 0, len(grid)-1) {
		return 0
	}
	if !inBounds(j, 0, len(grid[0])-1) {
		return 0
	}
	if grid[i][j] == '#' {
		return 1
	}
	return 0
}

func inBounds(a, l, r int) bool {
	return a >= l && a <= r
}

File: day11/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

type BErr = Box<dyn std::error::Error>;

fn main() -> Result<(), BErr> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut grid = Vec::new();
    let mut tmp_grid = Vec::new();
    for line in reader.lines() {
        let k = line?.chars().collect::<Vec<_>>();
        tmp_grid.push(vec!['\0'; k.len()]);
        grid.push(k);
    }
    let mut grid2 = grid.clone();
    let mut tmp_grid2 = tmp_grid.clone();

    until_stable(false, &mut grid, &mut tmp_grid);
    println!("Part 1: {}", count_seats(&grid));

    until_stable(true, &mut grid2, &mut tmp_grid2);
    println!("Part 2: {}", count_seats(&grid2));

    Ok(())
}

fn count_seats(grid: &Vec<Vec<char>>) -> usize {
    let mut count = 0;
    for i in grid {
        for &j in i {
            if j == '#' {
                count += 1;
            }
        }
    }
    count
}

fn until_stable<'a>(
    mode: bool,
    mut grid: &'a mut Vec<Vec<char>>,
    mut tmp_grid: &'a mut Vec<Vec<char>>,
) {
    loop {
        if !next_grid(mode, grid, tmp_grid) {
            break;
        }
        let tmp = grid;
        grid = tmp_grid;
        tmp_grid = tmp;
    }
}

fn next_grid(mode: bool, grid: &mut Vec<Vec<char>>, next: &mut Vec<Vec<char>>) -> bool {
    let mut change = false;
    for i in 0..grid.len() {
        for j in 0..grid[i].len() {
            let c = grid[i][j];
            let k = if mode {
                next_seat2(i, j, grid)
            } else {
                next_seat(i, j, grid)
            };
            if k != c {
                change = true;
            }
            next[i][j] = k;
        }
    }
    change
}

fn next_seat2(i: usize, j: usize, grid: &Vec<Vec<char>>) -> char {
    let c = grid[i][j];
    let s1 = seat_val2(i, j, 0, 1, 0, 0, grid);
    let s2 = seat_val2(i, j, 0, 1, 1, 0, grid);
    let s3 = seat_val2(i, j, 0, 0, 1, 0, grid);
    let s4 = seat_val2(i, j, 1, 0, 1, 0, grid);
    let s5 = seat_val2(i, j, 1, 0, 0, 0, grid);
    let s6 = seat_val2(i, j, 1, 0, 0, 1, grid);
    let s7 = seat_val2(i, j, 0, 0, 0, 1, grid);
    let s8 = seat_val2(i, j, 0, 1, 0, 1, grid);
    let k = s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8;
    if c == 'L' && k == 0 {
        return '#';
    }
    if c == '#' && k >= 5 {
        return 'L';
    }
    return c;
}

fn seat_val2(
    mut i: usize,
    mut j: usize,
    ui: usize,
    di: usize,
    uj: usize,
    dj: usize,
    grid: &Vec<Vec<char>>,
) -> usize {
    i = i + ui - di;
    j = j + uj - dj;
    while i < grid.len() && j < grid[0].len() {
        if grid[i][j] == 'L' {
            return 0;
        }
        if grid[i][j] == '#' {
            return 1;
        }
        i = i + ui - di;
        j = j + uj - dj;
    }
    0
}

fn next_seat(i: usize, j: usize, grid: &Vec<Vec<char>>) -> char {
    let c = grid[i][j];
    let s1 = seat_val(i - 1, j, grid);
    let s2 = seat_val(i - 1, j + 1, grid);
    let s3 = seat_val(i, j + 1, grid);
    let s4 = seat_val(i + 1, j + 1, grid);
    let s5 = seat_val(i + 1, j, grid);
    let s6 = seat_val(i + 1, j - 1, grid);
    let s7 = seat_val(i, j - 1, grid);
    let s8 = seat_val(i - 1, j - 1, grid);
    let k = s1 + s2 + s3 + s4 + s5 + s6 + s7 + s8;
    if c == 'L' && k == 0 {
        return '#';
    }
    if c == '#' && k >= 4 {
        return 'L';
    }
    return c;
}

fn seat_val(i: usize, j: usize, grid: &Vec<Vec<char>>) -> usize {
    if i >= grid.len() {
        return 0;
    }
    if j >= grid[0].len() {
        return 0;
    }
    if grid[i][j] == '#' {
        1
    } else {
        0
    }
}
